src/lib.rs

```diff
@@ -151,6 +161,45 @@ impl<I: DoubleEndedIterator> DoubleEndedIterator for PushBackIterator<I> {
     }
 }
 
+/// A lookahead iterator that doesn't consume the borrowed [`PushBackIterator`].
+///
+/// Unfortunately, we have to require [`Clone`] for the item type, because the
```

Comment:
It doesn't need to be clone. The requirement that a consumer needs to hold a reference and then advance the iterator is not needed for all consumers. Instead, this should return a reference and allow the consumer to decide whether or not to call `.clone()`

Reply:
I tried hard to make it so; my understanding is that it is impossible with the standard `Iterator` trait, because the prototype `next(&mut self) -> Self::Item` doesn't allow tying the lifetime of `&mut self` to the `Self::Item` type.

There are more details [here](http://lukaskalbertodt.github.io/2018/08/03/solving-the-generalized-streaming-iterator-problem-without-gats.html)

What can be done currently:

 - Implement a `fn next(&mut self) -> &I::Item` method outside of the `Iterator` trait. That only lets the user do `while Some(x) = i.next() {...}` 
 - Implement the trait from the `streaming-iterator` crate, as an optional dependency. It's not clear if `streaming-iterator` will hang around after GATs stabilize.
 - Implement the trait from `lending-iterator` as an optional dependency. That pulls `nougat` and some other things in the dependency tree.

Which direction do you prefer?

Reply:
I don't believe a streaming/lending iterator is needed for this, the type signature seems to match `slice::iter` which is able to be expressed using existing traits. https://doc.rust-lang.org/std/primitive.slice.html#method.iter. Perhaps I'm missing something.

Reply:
I may be wrong, but this is what I understood:

In the case of slices, the `Iter<'a, T>` object stores a `&'t [T]`. When iterated, this `&'t` can be copied out of the iterator, and then turned into an `&'[T]`, and all works well; both the `Iter` and the returned ref borrow from the original slice.

In our case, we need an `&'t mut PushbackIterator`. We can store it in the `LookaheadIterator`, but since &mut is not copy, we can't copy it out, so we have to mut borrow it from the iterator. So, we want the returned ref to borrow mut from the `LookaheadIterator`, which borrows mut from the underlying `PushbackIterator`; and this is what is unsupported by the Iterator trait.

I'd be happy to be proven wrong on this, but I tried really hard to satisfy the borrow checker to no avail, and I have no idea how to fix it.

Reply:
This naive implementation:

```rust
impl<'i, I: Iterator> Iterator for LookaheadIterator<'i, I>
where
    I::Item: Clone,
{
    type Item = &'i I::Item;

    fn next(&mut self) -> Option<Self::Item> {
        let next: Option<&'i I::Item> = self.inner.peek_nth(self.pos);
        self.pos += 1;
        next   
    }
}
```
results in:

```
error: lifetime may not live long enough
   --> src/lib.rs:199:19
    |
192 | impl<'i, I: Iterator> Iterator for LookaheadIterator<'i, I>
    |      -- lifetime `'i` defined here
...
198 |     fn next(&mut self) -> Option<Self::Item> {
    |             - let's call the lifetime of this reference `'1`
199 |         let next: Option<&'i I::Item> = self.inner.peek_nth(self.pos);
    |                   ^^^^^^^^^^^^^^^^^^^ type annotation requires that `'1` must outlive `'i`
```

I would like to write
```
fn next<'a: 'i>(&'a mut self) -> Option<Self::Item>
```

Then the borrow checker is happy, but:

`error[E0195]: lifetime parameters or bounds on method `next` do not match the trait declaration`